### PR TITLE
adds blueprint directory, template and example to solicit feedback

### PR DESCRIPTION
**Pull request by mitchell852**

### CODE REVIEW

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# *What* is being asked for?
+<!-- directions --->
+
+# *Why* is this necessary?
```
Comment (mitchell852) [inline blueprints/BLUEPRINT_TEMPLATE.md]: can you change this to "why is this needed? Are there any ways currently to accomplish this?

**Comment by rawlinp** [comment]
FYI I opened a PR to this PR here: https://github.com/mitchell852/trafficcontrol/pull/8

**Comment by lbathina** [comment]
I think the template is good. except that what level of info to be provided in each impact area.. say we should give in fields in specific and others - may be

**Comment by lbathina** [comment]
Also an example would give a picture of what is being proposing to the community

**Comment by mitchell852** [comment]
> Also an example would give a picture of what is being proposing to the community

there are 2 examples attached to this PR. one for the newly introduced client steering diversity feature and the pending tp columns configurability feature.

**Comment by mitchell852** [comment]
> I think the template is good. except that what level of info to be provided in each impact area.. say we should give in fields in specific and others - may be

Not really following. Do you have an example? Let's say TP is impacted. What kind of more granular information are you looking for?

**Comment by rob05c** [comment]
Can I suggest a heading, ideally near the top, of the status of a particular blueprint? I.e. proposed, accepted by the project for implementation, in progress (with links to PRs or in-master code), or implemented?

That way, people who haven't followed the mailing lists in detail can see whether it's just a proposal, or if the community has agreed to it and we're planning on implementing it?

**Comment by rawlinp** [comment]
I like that idea as well, but I kind of think that if the blueprint has been merged into the repo that should mean it's been accepted for implementation. Basically I think this would be process:
1. pitch basic idea to smaller set of interested parties to form the basis of a blueprint
2. write up blueprint
3. open blueprint PR against the project with `status: accepted`
4. send email the dev@ list to ask for feedback on the blueprint
5. any interested parties read the blueprint and review/add comments to the PR
6. once all feedback has been received/addressed on the PR, it will be merged
7. blueprint is then in the "accepted" state

Once it's in the "accepted" state, should you open a PR to the blueprint to set it to the "in-progress" state before starting its implementation? Then if you have one or more PRs to implement the blueprint, whatever is considered the "final" PR that finishes the blueprint would change the blueprint status to "complete". Once that "final" PR is merged, the blueprint would immediately reflect the `completed` status.

**Comment by rob05c** [comment]
>if the blueprint has been merged into the repo that should mean it's been accepted for implementation

I don't have a strong opinion on exactly how the process works, as long as it's clear. Which it wasn't, to me at least.

That also raises a few more questions: What about "blueprint" modifications? Do those also need approved by the list? To what degree? For example, does every tiny modification like changing an HTTP/JSON int to a float need a full vote on the list? IMO some rough guidelines on how big a change can be before it needs another vote would be good.

Without these, it won't be clear to new committers when a blueprint can be merged, when it's allowed to be worked on, etc.

How about a README.md in the `/blueprints` directory, with all of this info?

**Comment by mitchell852** [comment]
@rob05c 

> Can I suggest a heading, ideally near the top, of the status of a particular blueprint? I.e. proposed, accepted by the project for implementation, in progress (with links to PRs or in-master code), or implemented?

I feel like:

Blueprint PR is open == proposed
Blueprint PR is closed == rejected
Blueprint PR is merged == accepted

but maybe a status like:

[] in progress
[] implemented (include PR number)

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+How will existing endpoints be changed?
+What will the requests and responses look like?
+What fields are required or optional?
+What are the defaults for optional fields?
+What are the validation constraints?
+-->
+
+#### Client Impact
+<!--
+*How* will this impact Traffic Ops REST API clients (Go, Python, Java)?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:43]: If the REST API is changing, what answer is there other than YES.  If the rest API isn't changing what kind of work would you do that needs a blueprint?

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+If new endpoints are required, will corresponding client methods be added?
+-->
+
+#### Data Model / Database Impact
+<!--
+*How* will this impact the Traffic Ops data model?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:50]: How do you define the data model and database schema being different?

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+*How* will this impact the Traffic Ops database schema?
+
+What changes to the lib/go-tc structs will be required?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:53]: Isn't this part of the native client section?

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+What new tables and columns will be required?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:54]: Isn't this asked already when discussing schema changes?

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+### ORT Impact
+<!--
+*How* will this impact ORT?
+-->
+
+### Traffic Monitor Impact
+<!--
+*How* will this impact Traffic Monitor?
+
+Will new profile parameters be required?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:69]: Add question: Is new application configuration required that's not part of TO profile/params. (General statement for all ATC apps).

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+-->
+
+### Traffic Router Impact
+<!--
+*How* will this impact Traffic Router?
+
+Will new profile parameters be required?
+How will the CRConfig be changed?
+How will changes in Traffic Ops data be reflected in the CRConfig?
+Will Traffic Router remain backwards-compatible with old CRConfigs?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:79]: This feels like an immediate redflag.  It might be better to add a general ATC component question along the lines of `Are there any new ATC/ATS version dependencies introduced?`

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+What existing documentation will need to be updated?
+-->
+
+### Testing Impact
+<!--
+*How* will this impact testing?
+
+What is the high-level test plan?
+How should this be tested?
+Can this be tested within the existing test frameworks?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:110]: This could use a link to what existing test frameworks are.

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+*How* will this impact performance?
+
+Are the changes expected to improve performance in any way?
+Is there anything particularly CPU, network, or storage-intensive to be aware of?
+What are the known bottlenecks to be aware of that may need to be addressed?
+-->
+
+### Security Impact
+<!--
+*How* will this impact overall security?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:125]: This section needs the big disclaimers around the policy of reporting security vulnerabilities.

**Comment by ajschmidt** [comment]
I think its conceivable that a blueprint might become dependent on another feature which is in progress and has its own blueprint. I think we should either add another section referencing other blueprints or just add to the description of the existing dependencies section.

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+How will untrusted input be validated?
+If these changes are used maliciously or improperly, what could go wrong?
+Will these changes adhere to multi-tenancy?
+Will data be protected in transit (e.g. via HTTPS or TLS)?
+Will these changes require sensitive data that should be encrypted at rest?
+Will these changes require handling of any secrets?
+Will new SQL queries properly use parameter binding?
+-->
+
+### Upgrade Impact
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:139]: Instead of trying to lump these big topics (testing, operations, upgrades, etc) at the end, I think (personal preference) we should be asking those questions for every component with a simple overarching summary if needed.

File: blueprints/BLUEPRINT_TEMPLATE.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+Are there any security risks to be aware of?
+What privilege level is required for these changes?
+Do these changes increase the attack surface (e.g. new untrusted input)?
+How will untrusted input be validated?
+If these changes are used maliciously or improperly, what could go wrong?
+Will these changes adhere to multi-tenancy?
+Will data be protected in transit (e.g. via HTTPS or TLS)?
+Will these changes require sensitive data that should be encrypted at rest?
+Will these changes require handling of any secrets?
+Will new SQL queries properly use parameter binding?
```
Comment (jhg03a) [inline blueprints/BLUEPRINT_TEMPLATE.md:136]: Instead of asking leading questions where the answer should be YES, those kinds of things should go into a style guide.

**Comment by rawlinp** [comment]
I think that during the implementation of a blueprint if you realize that you have to change the blueprint (because you found a better way to implement something or have a better design), use your discretion in re-notifying the mailing list about the change to the blueprint. If you're changing a proposed data type for a new API from int to float, maybe the mailing list doesn't have to know, but there's no harm in sending a quick note to the list anyways. But if you're proposing a redesign of the initially-proposed API (e.g. going from /cachegroup_fallbacks to just adding a fallbacks array to the /cachegroups API object), the mailing list should probably be notified of that proposed change in the design. And I don't think the intention of the blueprint is really to upvote or downvote new changes -- it's about project communication and collaboration in the design process until the design satisfies most of the stakeholders (and if it _doesn't_ satisfy some stakeholders, those unsatisfactory trade-offs can be documented in their respective section of the blueprint).

I said this on the mailing list already, but I don't think anyone would really be "prohibited" from working on the implementation of a blueprint before it's been accepted. After all, the blueprint process itself would be optional (but encouraged). Although, it could end up saving a lot of time depending on how different the proposer's implementation idea is from what the final design ends up being after all interested stakeholders have participated. So maybe it would be _wise_ to wait until a blueprint has been merged before getting _too far_ into the implementation, but you wouldn't be _required_ to wait.

We could and probably should include these views in a README.md in the blueprints directory (assuming others agree with these views). I think I'll wait until the blueprint idea has been discussed more openly on the mailing list.

> I think its conceivable that a blueprint might become dependent on another feature which is in progress and has its own blueprint. I think we should either add another section referencing other blueprints or just add to the description of the existing dependencies section.

I think including that information in the existing `Dependencies` section would be fine. If you wanted to also link to the referenced blueprint you could include the URL in the `References` section like this:
```
## Dependencies
The implementation of this blueprint depends on the implementation of the [Foo blueprint] [1].

## References
[1]: https://example.com/foo-blueprint.md
```
That should embed a link in the "Foo blueprint" text that is also readable in the References section.

Feel free to open a PR against Jeremy's PR with another prompt for that in the Dependencies section if you want.

**Comment by rob05c** [comment]
>I feel like:
>Blueprint PR is open == proposed
>Blueprint PR is closed == rejected
>Blueprint PR is merged == accepted

@mitchell852 No objections, as long as that's documented in /blueprints/README.md

**Comment by ocket8888** [comment]
This has been adopted upstream, example no longer necessary
